pyvrp/cpp/DurationSegment.h

```diff
@@ -187,10 +187,14 @@ Duration DurationSegment::duration() const { return duration_; }
 
 Duration DurationSegment::timeWarp(Duration const maxDuration) const
 {
+    // We apply the maximum duration constraint to the net duration, which is
+    // the difference between the start and end time of the route which equals
+    // the total duration minus existing time warp.
+    // Hence, if duration - timeWarp > maxDuration, we increase timeWarp such
+    // duration - timeWarp = maxDuration.
     // clang-format off
-    return timeWarp_
-         + std::max<Duration>(releaseTime_ - twLate_, 0)
-         + std::max<Duration>(duration_ - maxDuration, 0);
+    return std::max<Duration>(duration_ - maxDuration, timeWarp_)
+         + std::max<Duration>(releaseTime_ - twLate_, 0);
```

Comment:
An alternative would be to also consider timewarp for release times as existing time warp. It depends a bit on what we define exactly as the start of the route and the duration which we want to compare against maximum duration in this case. For now I chose to leave it as is.
```suggestion
    return std::max<Duration>(duration_ - maxDuration, timeWarp_
         + std::max<Duration>(releaseTime_ - twLate_, 0));
```